### DISCUSSION TITLE
How does autoload of plugins work?

### DISCUSSION THREAD

**Discussion by Stefanhg** [pytest-dev/pytest#11615] [Q&A] [answered]
Hello,
I am debugging on a issue where my plugin is loaded even when not included in the `pytest_plugins` list.
In order to run my test I have to `pip install .` else it does not load it.
After this when running the test the plugin I am working on is loaded even when `pytest_plugins` only includes pytester

Any idea why so? And can somebody explain how far the 'autoload' of plugins goes? Does it actually load every hook in the plugin or only registers it in PyTest?

**Comment by RonnyPfannschmidt** [comment, marked as answer]
Pytest instructs pluggy to load and register all plugins declared by a entry point

Entry points are managed in the package metadata, which is why a install/editable install is necessary

**Reply by Stefanhg** [reply]
Alright. So the autoload does only attach the plugin, and PyTest then hooks it if included in `pytest_plugins`?

Noted!

**Reply by RonnyPfannschmidt** [reply]
No, any Plugin loaded by a entry point is enabled in full

**Reply by Stefanhg** [reply]
Well that is not smart imho. What is the function of `pytest_plugins` then? 

So that means it is expected doing testing of a plugin to have it execute twice and the only way to disable a single plugin is to uninstall it or enable the PYTEST_DISABLE_PLUGIN_AUTOLOAD then I would assume `pytest_plugins` would be the source of activated plugins.

**Reply by Stefanhg** [reply]
My assumptions are incorrect it seems. With `PYTEST_DISABLE_PLUGIN_AUTOLOAD `the plugin in `pytest_plugins` NOT loaded. 
So when a plugin is installed you have no control of it being in use or not. 

I assume the reason for autoload is that it is much easier to manage if you have many plugins installed but I would also assume that you know exactly what plugins is in use.
But It could also become messy if then another plugin relies on a different plugin, that means the plugins has to have support for including `pytest_plugins`.

**Reply by RonnyPfannschmidt** [reply]
It's not quite clear what you are asking for
